### PR TITLE
update limiter module

### PR DESCRIPTION
**Pull request by MouceL**
support new smartlimter crds
support single/global strategy
support e2e

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
 FROM ubuntu:bionic
 WORKDIR /
-COPY manager.exe ./manager
+COPY manager ./manager
```
Comment (YonkaFang) [inline Dockerfile:5]: leave the .exe ext to apply git ignore

File: VERSION [hunk 2]
```diff
@@ -0,0 +1 @@
+v0.2.3
```
Comment (YonkaFang) [inline VERSION:1]: VERSION file is not needed by separated module which uses git tag as version.

File: controllers/envoy_global_limit.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+// https://github.com/envoyproxy/ratelimit only support per second, minute, hour, and day limits
+func calculateQuotaPerUnit(descriptor *microservicev1alpha1.SmartLimitDescriptor) (quota int, unit string, err error) {
+	quota, err = strconv.Atoi(descriptor.Action.Quota)
+	if err != nil {
+		return quota, unit, err
+	}
+	seconds := descriptor.Action.FillInterval.Seconds
+	switch seconds {
+	case 60 * 60 * 24:
```
[the remark this answers is not in the document]
Comment (YonkaFang) [reply controllers/envoy_global_limit.go:135]: and commented

File: controllers/envoy_filters.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	} else if len(httpRouterPatches) > 0 {
+		ef.ConfigPatches = append(ef.ConfigPatches, httpRouterPatches...)
+	}
+
+	// config plugin envoy.filters.http.ratelimit
+	if len(globalDescriptors) > 0 {
+		var rlServer string
+		for _, item := range globalDescriptors {
+			if item.Action.RateLimitService != "" {
+				rlServer = item.Action.RateLimitService
```
[the remark this answers is not in the document]
Comment (MouceL) [reply controllers/envoy_filters.go:126]: > what if these `RateLimitService`s diff?

it's follow the first now ,  or i can exact the this config to other places

File: controllers/envoy_filters.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		server := getRateLimiterServerCluster(rlServer)
+		httpFilterEnvoyRateLimitPatch := generateEnvoyHttpFilterGlobalRateLimitPatch(server)
+		if httpFilterEnvoyRateLimitPatch != nil {
+			ef.ConfigPatches = append(ef.ConfigPatches, httpFilterEnvoyRateLimitPatch)
+		}
+	}
+
+	// enable and config plugin envoy.filters.http.local_ratelimit
+	if len(localDescriptors) > 0 {
+		httpFilterLocalRateLimitPatch := generateHttpFilterLocalRateLimitPatch()
```
[the remark this answers is not in the document]
Comment (MouceL) [reply controllers/envoy_filters.go:139]: > I'm not sure if the enable-ratelimit-filter thing should be done by us here as there maybe an exsiting one and then we'll get dup rate limit filters. It's really difficult to deal with. Maybe istio should provide a new `Operation` `INSERT_BEFORE_IF_NOT_EXIST`.

it's a question, we don't know the existene of the plugins before insert

File: publish.sh [hunk 6]
```diff
@@ -1,4 +1,4 @@
 #!/usr/bin/env bash
 
 export MOD=limiter
-../slime/bin/publish.sh "$@"
+../../bin/publish.sh "$@"
```
Comment (YonkaFang) [inline publish.sh:4]: we should respect the file hierarchy convention:
```sh
<your_module>/
slime/
```

File: controllers/smartlimiter_controller.go [hunk 7]
```diff
@@ -82,9 +80,16 @@ func (r *SmartLimiterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
...
+		r.lastUpdatePolicy = microservicev1alpha2.SmartLimiterSpec{}
 		r.lastUpdatePolicyLock.Unlock()
-		return reconcile.Result{}, nil
+		//if contain global smart limiter, should delete info in configmap
+		refreshConfigMap([]*model.Descriptor{}, r, req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	if len(instance.Spec.Sets) == 0 {
+		log.Infof("sets is nil,continue")
```
Comment (YonkaFang) [inline controllers/smartlimiter_controller.go:91]: no reconcile?

File: controllers/smartlimiter_controller.go [hunk 8]
```diff
@@ -66,9 +66,7 @@ type SmartLimiterReconciler struct {
 func (r *SmartLimiterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	_ = context.Background()
 
-	// your logic here
-
-	instance := &microserviceslimeiov1alpha1.SmartLimiter{}
+	instance := &microservicev1alpha2.SmartLimiter{}
```
Comment (YonkaFang) [inline controllers/smartlimiter_controller.go:69]: rev support code lost

File: controllers/envoy_filters.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+	} else {
+		sets = make([]*networking.Subset, 0, 1)
+	}
+	sets = append(sets, &networking.Subset{Name: util.Wellkonw_BaseSet})
+
+	loc := types.NamespacedName{Namespace: instance.Namespace, Name: instance.Name}
+	svc := &v1.Service{}
+	if err := r.Client.Get(context.TODO(), loc, svc); err != nil {
+		if errors.IsNotFound(err) {
+			log.Errorf("svc %s:%s is not found", loc.Name, loc.Namespace)
```
[the remark this answers is not in the document]
Comment (MouceL) [reply controllers/envoy_filters.go:38]: return when svc is missing

**Pull request merged**